<a href="https://colab.research.google.com/github/dev02chandan/recommendation-systems/blob/main/Singular%20Value%20Decomposition.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Recommendation Using Singular Value Decomposition**

1.	Create a toy dataset
2.	Split the data into train and test (70:30)
3.	Applying SVD to Toy matrix A

    a.	Testing using SVD in Toy dataset Selecting k = 1,2,3 and 5. Which value of k gives the best RMSE? What is the relationship between the k and RMSE?

    b.	Test with New user with rating:  1 1 3 0 4

4.	On Movie Lense dataset(Ratings.csv)

    a.	Data Exploration

    b.	Model Creation without SVD

    c.	RMSE without SVD  (Report the best RMSE by tuning the hyper parameters of the model)

    d.	Model creation using SVD

    e.	Model tuning for different value of k

    f.	RMSE using SVD

    g.	Generate top-10 recommendations for a target user.




## Task 1,2,3

In [1]:
import numpy as np

In [2]:
user_item_matrix = np.random.randint(low = 1, high = 6, size=(10, 10))

In [3]:
user_item_matrix

array([[3, 3, 5, 4, 3, 5, 3, 2, 3, 4],
       [3, 5, 1, 1, 3, 3, 2, 2, 5, 1],
       [3, 2, 1, 5, 4, 4, 2, 3, 1, 3],
       [4, 1, 4, 2, 2, 5, 4, 3, 2, 1],
       [3, 2, 2, 1, 3, 3, 1, 1, 5, 4],
       [2, 4, 4, 1, 4, 5, 5, 3, 1, 1],
       [3, 2, 2, 2, 2, 1, 3, 1, 4, 5],
       [5, 5, 1, 2, 1, 2, 2, 2, 5, 4],
       [2, 3, 1, 3, 5, 3, 3, 5, 3, 2],
       [2, 5, 4, 3, 1, 2, 2, 3, 4, 3]])

In [4]:
from sklearn.model_selection import train_test_split

In [5]:
train_ds, test_ds = train_test_split(user_item_matrix, test_size=0.3 )

In [6]:
train_ds

array([[4, 1, 4, 2, 2, 5, 4, 3, 2, 1],
       [2, 3, 1, 3, 5, 3, 3, 5, 3, 2],
       [2, 4, 4, 1, 4, 5, 5, 3, 1, 1],
       [5, 5, 1, 2, 1, 2, 2, 2, 5, 4],
       [3, 2, 2, 2, 2, 1, 3, 1, 4, 5],
       [2, 5, 4, 3, 1, 2, 2, 3, 4, 3],
       [3, 5, 1, 1, 3, 3, 2, 2, 5, 1]])

In [7]:
test_ds

array([[3, 3, 5, 4, 3, 5, 3, 2, 3, 4],
       [3, 2, 1, 5, 4, 4, 2, 3, 1, 3],
       [3, 2, 2, 1, 3, 3, 1, 1, 5, 4]])

In [8]:
#Computing Total Mean Squared Error
#--------------------------------
from numpy.linalg import norm

def error(R,P,Q,lamda=0.02):
    ratings = R.data
    rows = R.row
    cols = R.col
    e = 0
    for ui in range(len(ratings)):
        rui=ratings[ui]
        u = rows[ui]
        i = cols[ui]
        if rui>0:
            e= e + pow(rui-np.dot(P[u,:],Q[:,i]),2)+\
                lamda*(pow(norm(P[u,:]),2)+pow(norm(Q[:,i]),2))
    return e

In [9]:
M, N = train_ds.shape

In [10]:
M

7

In [11]:
N

10

In [12]:
K = [1, 2, 3, 5]

In [13]:
P = np.random.rand(M, K[2])
Q = np.random.rand(K[2], N)

In [14]:
P

array([[0.18169571, 0.65659481, 0.3096181 ],
       [0.59480881, 0.48021584, 0.01668502],
       [0.64610541, 0.2856222 , 0.16476582],
       [0.49775885, 0.06175918, 0.53166375],
       [0.7816273 , 0.38988841, 0.61573954],
       [0.6432536 , 0.54820354, 0.83015728],
       [0.06186277, 0.72988701, 0.19683531]])

In [15]:
Q

array([[0.2783167 , 0.44957434, 0.58848332, 0.33638678, 0.61875182,
        0.93816148, 0.72372275, 0.35430417, 0.64310911, 0.56224058],
       [0.79240447, 0.41433745, 0.57165187, 0.10660029, 0.87387431,
        0.92539814, 0.06388404, 0.3390655 , 0.97872873, 0.53788045],
       [0.653182  , 0.99109022, 0.06074422, 0.61966698, 0.69540809,
        0.42728788, 0.4927993 , 0.30095739, 0.26040443, 0.73193088]])

In [16]:
from scipy.sparse import coo_matrix

R = coo_matrix(train_ds)

print ("R Shape::", R.shape)
print ("R Columns::", R.col)
print ("R Rows::",R.row)

R Shape:: (7, 10)
R Columns:: [0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6
 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9]
R Rows:: [0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 3 3 3 3 3 3 3
 3 3 3 4 4 4 4 4 4 4 4 4 4 5 5 5 5 5 5 5 5 5 5 6 6 6 6 6 6 6 6 6 6]


In [17]:
error(R, P, Q)

442.37399346348633

In [18]:
for i in K:
    M, N = train_ds.shape
    P = np.random.rand(M, i)
    Q = np.random.rand(i, N)
    R = coo_matrix(train_ds)
    print("K = ", i, " Error: ", error(R, P, Q))

K =  1  Error:  607.1998393841434
K =  2  Error:  502.1049831518942
K =  3  Error:  450.9473010972853
K =  5  Error:  282.984663734113


**Observation:** As the value of K increases (1, 2, 3, 5), the Error decreases

In [19]:
import scipy

In [20]:
U, S, Vh = scipy.linalg.svd(train_ds, full_matrices=True)

In [21]:
U.shape, S.shape, Vh.shape

((7, 7), (7,), (10, 10))

In [22]:
Uk = U[:, :K[2]]

In [23]:
Sk = np.diag(S[:K[2]])

In [24]:
Vhk = Vh[:K[2], :]

In [25]:
Uk.shape, Sk.shape, Vhk.shape

((7, 3), (3, 3), (3, 10))

In [26]:
Uk @ Sk @ Vhk

array([[3.00495558, 1.77523981, 4.46122863, 1.80674204, 2.29741865,
        4.47652767, 4.43752139, 2.68003928, 1.42409389, 1.74560546],
       [2.13141485, 4.42789141, 1.52830165, 1.88479421, 4.34940558,
        3.61258732, 2.96896525, 3.90581465, 3.21008162, 0.998453  ],
       [2.43989129, 2.76560078, 3.72564357, 1.83033087, 3.96668045,
        5.07471065, 4.50900934, 3.82484873, 1.55099594, 0.72437254],
       [3.92106472, 4.68099131, 1.68306593, 2.39539775, 1.45146192,
        1.5073765 , 2.02227298, 1.98700983, 5.44476643, 4.29640623],
       [3.55114581, 3.04201196, 2.40393071, 1.97840218, 0.60722039,
        1.61750506, 2.22468765, 1.31135481, 3.92551041, 3.78473469],
       [3.48742884, 3.85840108, 2.38039465, 2.17505021, 1.91914268,
        2.45835383, 2.72038732, 2.32245924, 4.14692407, 3.29745233],
       [2.48628164, 4.66639824, 0.87647988, 1.93917406, 3.21902597,
        2.28621493, 2.04142464, 3.041408  , 4.17411174, 2.07900662]])

In [27]:
reconstructed = Uk @ Sk @ Vhk

In [28]:
rmse = np.sqrt(np.mean((train_ds - reconstructed) ** 2))

In [29]:
rmse

0.7191161746692761

In [30]:
for i in K:
    U, S, Vh = scipy.linalg.svd(train_ds, full_matrices=True)
    Uk = U[: , :i]
    Sk = np.diag(S[:i])
    Vhk = Vh[:i, :]
    reconstructed = Uk @ Sk @ Vhk
    rmse = np.sqrt(np.mean((train_ds - reconstructed) ** 2))
    print("K = ", i, " RSMError: ", rmse)

K =  1  RSMError:  1.2741816280118967
K =  2  RSMError:  0.9067897486177955
K =  3  RSMError:  0.7191161746692761
K =  5  RSMError:  0.36744638451205247


**Observation:** As the value of K increases (1, 2, 3, 5), the Error decreases

In [31]:
new_user_ratings = [1, 1, 3, 0, 4]
new = np.array(new_user_ratings)

In [32]:
new

array([1, 1, 3, 0, 4])

In [33]:
user_item_matrix = np.random.randint(low = 1, high = 6, size=(5,5))
user_item_matrix

array([[3, 1, 5, 3, 1],
       [2, 2, 4, 2, 5],
       [5, 2, 2, 1, 5],
       [1, 2, 4, 2, 2],
       [3, 5, 1, 1, 1]])

In [34]:
user_item_matrix = np.append(user_item_matrix, [new], axis=0)
user_item_matrix

array([[3, 1, 5, 3, 1],
       [2, 2, 4, 2, 5],
       [5, 2, 2, 1, 5],
       [1, 2, 4, 2, 2],
       [3, 5, 1, 1, 1],
       [1, 1, 3, 0, 4]])

In [35]:
U, S, Vh = scipy.linalg.svd(user_item_matrix, full_matrices=True)
Uk = U[: , :3]
Sk = np.diag(S[:3])
Vhk = Vh[:3, :]
reconstructed = Uk @ Sk @ Vhk
reconstructed

array([[1.95145597, 1.8100656 , 5.07909994, 2.93463331, 1.28739213],
       [2.69862908, 1.54276025, 4.16544944, 1.65111867, 4.7264195 ],
       [3.83714877, 2.93905324, 2.1952942 , 0.7339978 , 5.2782409 ],
       [1.74934683, 1.41073763, 3.91611345, 2.09592779, 1.80490907],
       [3.66076884, 4.45047514, 0.84691462, 1.22690792, 0.85774276],
       [1.56915407, 0.42155706, 2.59018506, 0.6954595 , 3.98206682]])

After SVD, the predicted value is negative depicting the limitation of SVD. SVD output ratings are not bound in a certain range.

## Reading the Dataset and Exploratory Data analysis

In [36]:
import pandas as pd

In [37]:
df = pd.read_csv('/content/drive/MyDrive/RS/4ratings.csv')
df

userId  movieId  rating   timestamp
0            1        1     4.0   964982703
1            1        3     4.0   964981247
2            1        6     4.0   964982224
3            1       47     5.0   964983815
4            1       50     5.0   964982931
...        ...      ...     ...         ...
100831     610   166534     4.0  1493848402
100832     610   168248     5.0  1493850091
100833     610   168250     5.0  1494273047
100834     610   168252     5.0  1493846352
100835     610   170875     3.0  1493846415

[100836 rows x 4 columns]

In [38]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100836 entries, 0 to 100835
Data columns (total 4 columns):
 #   Column     Non-Null Count   Dtype  
---  ------     --------------   -----  
 0   userId     100836 non-null  int64  
 1   movieId    100836 non-null  int64  
 2   rating     100836 non-null  float64
 3   timestamp  100836 non-null  int64  
dtypes: float64(1), int64(3)
memory usage: 3.1 MB


In [41]:
# Number of users
print("The ratings dataset has ", df['userId'].nunique(), "unique users")

# Number of movies
print("The ratings dataset has ", df['movieId'].nunique(), "unique movies")

# Number of ratings
print("The ratings dataset has ", df['rating'].nunique(), "unique ratings")

# List of unique ratings
print("the unique ratings are ", sorted(df['rating'].unique()))

The ratings dataset has  610 unique users
The ratings dataset has  9724 unique movies
The ratings dataset has  10 unique ratings
the unique ratings are  [0.5, 1.0, 1.5, 2.0, 2.5, 3.0, 3.5, 4.0, 4.5, 5.0]


In [42]:
df.head()

userId  movieId  rating  timestamp
0     1       1       4.0   964982703
1     1       3       4.0   964981247
2     1       6       4.0   964982224
3     1      47       5.0   964983815
4     1      50       5.0   964982931

In [43]:
# Calculate the characteristics of the dataset
total_ratings = len(df)
unique_users = df['userId'].nunique()
unique_movies = df['movieId'].nunique()
rating_distribution = df['rating'].value_counts().sort_index()
avg_ratings_per_user = total_ratings / unique_users
avg_ratings_per_movie = total_ratings / unique_movies

total_ratings, unique_users, unique_movies, rating_distribution, avg_ratings_per_user, avg_ratings_per_movie


(100836,
 610,
 9724,
 0.5     1370
 1.0     2811
 1.5     1791
 2.0     7551
 2.5     5550
 3.0    20047
 3.5    13136
 4.0    26818
 4.5     8551
 5.0    13211
 Name: rating, dtype: int64,
 165.30491803278687,
 10.369806663924312)

**Observations**

Total Ratings: 100,836

Unique Users: 610

Unique Movies: 9,724

Distribution of Ratings:

0.5: 1,370

1.0: 2,811

1.5: 1,791

2.0: 7,551

2.5: 5,550

3.0: 20,047

3.5: 13,136

4.0: 26,818

4.5: 8,551

5.0: 13,211

Average Number of Ratings per User: ~165.3

Average Number of Ratings per Movie: ~10.4

## Model Creation without SVD

In [44]:
!pip install scikit-surprise # provides tools for building and analyzing recommendation systems

In [45]:
from surprise import Dataset, Reader
from surprise.model_selection import train_test_split

reader = Reader(rating_scale=(0.5, 5.0))
data = Dataset.load_from_df(df[['userId', 'movieId', 'rating']], reader)
trainset, testset = train_test_split(data, test_size=0.25)

In [46]:
from surprise import KNNBasic, accuracy

user_based_collab_filter = KNNBasic(sim_options={'user_based': True})
user_based_collab_filter.fit(trainset)
predictions = user_based_collab_filter.test(testset)

Computing the msd similarity matrix...
Done computing similarity matrix.


In [47]:
rmse_user_based = accuracy.rmse(predictions)
print(rmse_user_based)

RMSE: 0.9500
0.9499781500653749


## RMSE without SVD
(Report the best RMSE by tuning the hyper parameters of the model)

In [48]:
from surprise.model_selection import GridSearchCV

param_grid = {
    'k': [20, 40, 60, 80, 100],  # max number of neighbors
    'sim_options': {
        'name': ['msd', 'cosine', 'pearson'],
        'user_based': [True],
    }
}

gs = GridSearchCV(KNNBasic, param_grid, measures=['rmse'], cv=3)
gs.fit(data)

print(gs.best_score['rmse'])
print(gs.best_params['rmse'])

Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the cosine similarity matrix...
Done computing similarity matrix.
Computing the cosine similarity matrix...
Done computing similarity matrix.
Computing the cosine similarity matrix...
Done computing similarity matrix.
Computing the pearson similarity matrix...
Done computing similarity matrix.
Computing the pearson similarity matrix...
Done computing similarity matrix.
Computing the pearson similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the cosine similarity matrix...
Done computing similarity matrix.
Computing the cosine simila

## Model creation using SVD

In [49]:
from surprise import SVD

svd_model = SVD()
svd_model.fit(trainset)
svd_predictions = svd_model.test(testset)

In [50]:
rmse_svd = accuracy.rmse(svd_predictions)
print(rmse_svd)

RMSE: 0.8767
0.876723522134192


## Model Tuning for different values of *K*

In [51]:
param_grid = {
    'n_factors': [50, 100, 150, 200],
    'lr_all': [0.002, 0.005],
    'reg_all': [0.02, 0.1]
}

gs_svd = GridSearchCV(SVD, param_grid, measures=['rmse'], cv=3)
gs_svd.fit(data)

print(gs_svd.best_score['rmse'])
print(gs_svd.best_params['rmse'])

0.875569999875002
{'n_factors': 50, 'lr_all': 0.005, 'reg_all': 0.1}


## **Generate top-10 recommendations for a target user**

In [52]:
def get_top_n_recommendations_with_actual(predictions, n=10):
    # Map the predictions to each user
    top_n = {}
    for uid, iid, true_r, est, _ in predictions:
        top_n.setdefault(uid, []).append((iid, true_r, est))

    # Sort the predictions for each user and retrieve the n highest ones
    for uid, user_ratings in top_n.items():
        user_ratings.sort(key=lambda x: x[2], reverse=True)  # Sort by estimated rating
        top_n[uid] = user_ratings[:n]

    return top_n

In [53]:
top_recommendations_with_actual = get_top_n_recommendations_with_actual(svd_predictions, n=10)

print('Movie Id \t Actual Rating \t Predicted Rating')
target_user = 1
for i in top_recommendations_with_actual[target_user]:
    print(i[0], '\t\t', i[1], '\t\t', i[2])

print("The top n predicted ratings were recommended to the user.")

Movie Id 	 Actual Rating 	 Predicted Rating
260 		 5.0 		 5.0
1291 		 5.0 		 5.0
1219 		 2.0 		 4.9863518986808835
1198 		 5.0 		 4.919494175627743
3703 		 5.0 		 4.903227981045364
2858 		 5.0 		 4.890779540000871
2916 		 4.0 		 4.796825896108422
1954 		 5.0 		 4.793032071194854
1220 		 5.0 		 4.7841164855544775
2329 		 5.0 		 4.783894627363631
The top n predicted ratings were recommended to the user.


# **Conclusion**

First we did SVD on a toy dataset.
We observed that as the value of K increases, the RMSE also decreases. In our exploration of the MovieLens dataset, we delved into recommendation system methodologies, starting with user-based collaborative filtering and progressing to matrix factorization via Singular Value Decomposition (SVD). While collaborative filtering leverages user similarities to predict preferences, SVD uncovers latent factors that drive user-movie interactions. Evaluating models using the RMSE metric highlighted their predictive accuracy. Visualization efforts, though faced with technical challenges, aimed to offer insights into user rating behaviors and patterns. Ultimately, the endeavor showcased the potential of data-driven techniques in curating personalized movie recommendations, emphasizing the synergy between data exploration, modeling, evaluation, and user-centric output.